src/ui/Editor.cpp

```diff
@@ -1395,7 +1396,65 @@ Editor::GetSymbol()
 	int32 size = end - start;
 	char text[size];
 	GetText(start, size, text);
-	return text;
+	// remove invalid leading characters
+	const std::regex leadingChars("^\\W+");
+	std::string str = std::regex_replace(text, leadingChars, "");
+	return BString(str.c_str());
+}
+
+
+void
+Editor::SetSelection(int32 start, int32 end)
+{
+	SendMessage(SCI_SETSELECTIONSTART, start);
+	SendMessage(SCI_SETSELECTIONEND, end);
+}
+
+
+void
+Editor::Append(BString text)
+{
+	SendMessage(SCI_BEGINUNDOACTION, 0, 0);
+	SendMessage(SCI_APPENDTEXT, text.Length(), (sptr_t)text.String());
+	ScrollCaret();
+	SendMessage(SCI_ENDUNDOACTION, 0, 0);
+}
+
+
+void
+Editor::Insert(BString text, int32 start)
+{
+	SendMessage(SCI_BEGINUNDOACTION, 0, 0);
+	if (start == -1)
+		SendMessage(SCI_REPLACESEL, UNSET, (sptr_t)text.String());
+	else
+		SendMessage(SCI_INSERTTEXT, start, (sptr_t)text.String());
+	ScrollCaret();
+	SendMessage(SCI_ENDUNDOACTION, 0, 0);
+}
+
+
+const BString
+Editor::GetLine(int32 lineNumber)
+{
+	int32 lineLength = SendMessage(SCI_LINELENGTH, lineNumber, UNSET);
+	char *lineBuffer = new char[lineLength + 1];
+	memset(lineBuffer, '\0', lineLength + 1);
```

Comment:
lineBuffer[lineLength]='\0' should be enough

Reply:
done